src/sources/redis/stream.rs

```diff
@@ -0,0 +1,113 @@
+use std::str::from_utf8;
+
+use crate::{
+    codecs,
+    config::SourceContext,
+    internal_events::RedisReceiveEventError,
+    sources::{redis::handle_line, Source},
+};
+use redis::{aio::Connection, streams::*, AsyncCommands, RedisError, RedisResult, Value};
+use snafu::{ResultExt, Snafu};
+use vector_common::internal_event::{BytesReceived, Registered};
+use vector_core::config::LogNamespace;
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("Failed to create connection: {}", source))]
+    Connection { source: RedisError },
+}
+
+pub struct StreamInputs {
+    pub client: redis::Client,
+    pub bytes_received: Registered<BytesReceived>,
+    pub key: String,
+    pub redis_key: Option<String>,
+    pub decoder: codecs::Decoder,
+    pub cx: SourceContext,
+    pub log_namespace: LogNamespace,
+}
+
+fn redis_value_to_string(value: &Value) -> String {
+    let mut result_string = String::new();
+
+    match value {
+        redis::Value::Nil => {}
+        redis::Value::Int(number) => result_string.push_str(&number.to_string()),
+        redis::Value::Data(byte_array) => result_string.push_str(from_utf8(byte_array).unwrap()),
+        redis::Value::Bulk(values) => {
+            for val in values.iter() {
+                result_string.push_str(redis_value_to_string(val).as_str());
+                result_string.push(' ');
+            }
+        }
+        redis::Value::Status(status) => result_string.push_str(status),
+        redis::Value::Okay => result_string.push_str("OK"),
+    }
+
+    result_string
+}
+
+fn stream_id_to_string(stream_id: StreamId) -> String {
+    let mut result_string = String::new();
+
+    for (key, value) in stream_id.map.iter() {
+        result_string.push_str(key);
+        result_string.push(' ');
+        result_string.push_str(redis_value_to_string(value).as_str());
+        result_string.push(' ');
+    }
+
+    result_string
+}
+
+async fn read_stream(mut conn: Connection, key: String) -> RedisResult<String> {
+    let opts = StreamReadOptions::default().count(1).block(0);
```

Comment:
We don't need to read just one event at a time. This should read as many events as necessary and create a `Vec<Event>` to send to the output.